### DISCUSSION TITLE
Is it possible to change lag weights of ad-stocking in custom prior at a individual channel level?

### DISCUSSION THREAD

**Discussion by AruunnAnanda** [google/lightweight_mmm#178] [Q&A]
Hi there,

As of now, I'm using custom prior to reduce my lag weight of ad-stocking. And this custom prior tweak **applies to all channels**. Below is my code for reference. 

I would like to individually modify lag weights for specific channels using custom prior. Is there are a way already to do it? or is this something not achievable? Looking forward to your help! Thanks! 

`mmm_hill_adstock.fit(
    media=media_data_train,
    media_prior=costs,
    target=target_train,
    extra_features=extra_features_train,
    number_warmup=number_warmup,
    number_samples=number_samples,
    seed=SEED,
    media_names=['brand_s', 'other_display_s', 'liftoff_s',
       'sem_non_branded_s', 'sem_branded_s', 'sem_google_uac_s',
       'paid_social_s','parcel_s', 'podcast_s', 'tv_s', 'ooh_s', 'radio_s'],
    custom_priors={"lag_weight": numpyro.distributions.Beta(concentration1=1, concentration0=4)})`

**Comment by steven-struglia** [comment]
Regarding changing the `lag_weight` parameter in `custom_priors` on an individual level, you can do it like this:

```
c1 = np.array([2] * len(media_columns))
c1[7] = 1.5
c0 = np.ones(len(media_columns))
custom_priors = {'lag_weight': {'concentration1': c1, 'concentration0': c0}}
```
Note that essentially you're creating 2 arrays of concentrations to feed into the Beta distribution, each of length `len(media_names)` where the indexes are kept the same. In the example above, I'm changing the 7th index media channel to use a `concentration1 = 1.5` instead of 2.

Keep in mind that the default distribution for `lag_weight` is a `dist.Beta(concentration1=2., concentration0=1.)`. If you want to change it to a different distribution, you must change the default distribution in `models.py` line 132 and refresh the package, python kernel, and re-import the modified package. I have personally changed mine to be a Uniform distribution with custom priors in order to have a weaker prior and more accurate results across the board according to my intuition and theoretical knowledge on MMM's.

**Reply by tbichu** [reply]
Hi @steven-struglia thanks for the comment.
I tried this approach a few months back but now it seems to throw the following error when I try to pass individual priors for each channel using this :
custom_priors = {'lag_weight': {'concentration1': c1.astype(float), 'concentration0': c0.astype(float)}, 'exponent': {'concentration1': e1.astype(float), 'concentration0': e1.astype(float)}}

However if I just pass 
custom_priors = {'lag_weight': {'concentration1': 2.0, 'concentration0': 2.0}, 'exponent': {'concentration1': 4.0, 'concentration0': 1.0}}
then it works. 
Not sure what might be happening here.


`---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
/tmp/ipykernel_517046/291371203.py in <cell line: 50>()
     48 
     49 mmm_adstock = lightweight_mmm.LightweightMMM(model_name="adstock")
---> 50 mmm_adstock.fit(media = media_data_train, media_prior=costs, target=target_train, media_names = spends_cols, seed=SEED, seasonality_frequency = 52, custom_priors = custom_priors2)
     51 # custom_priors = custom_priors

/opt/conda/miniconda3/lib/python3.10/site-packages/lightweight_mmm/lightweight_mmm.py in fit(self, media, media_prior, target, extra_features, degrees_seasonality, seasonality_frequency, weekday_seasonality, media_names, number_warmup, number_samples, number_chains, target_accept_prob, init_strategy, custom_priors, seed)
    362         num_samples=number_samples,
    363         num_chains=number_chains)
--> 364     mcmc.run(
    365         rng_key=jax.random.PRNGKey(seed),
    366         media_data=jnp.array(media),

/opt/conda/miniconda3/lib/python3.10/site-packages/numpyro/infer/mcmc.py in run(self, rng_key, extra_fields, init_params, *args, **kwargs)
    561         self._args = args
    562         self._kwargs = kwargs
--> 563         init_state = self._get_cached_init_state(rng_key, args, kwargs)
    564         if self.num_chains > 1 and rng_key.ndim == 1:
    565             rng_key = random.split(rng_key, self.num_chains)

/opt/conda/miniconda3/lib/python3.10/site-packages/numpyro/infer/mcmc.py in _get_cached_init_state(self, rng_key, args, kwargs)
    368         rng_key = (_hashable(rng_key),)
    369         args = tree_map(lambda x: _hashable(x), args)
--> 370         kwargs = tree_map(lambda x: _hashable(x), tuple(sorted(kwargs.items())))
    371         key = rng_key + args + kwargs
    372         try:

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/tree_util.py in tree_map(f, tree, is_leaf, *rest)
    242   leaves, treedef = tree_flatten(tree, is_leaf)
    243   all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
--> 244   return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
    245 
    246 

/opt/conda/miniconda3/lib/python3.10/site-packages/numpyro/distributions/distribution.py in tree_unflatten(cls, aux_data, params)
    149     @classmethod
    150     def tree_unflatten(cls, aux_data, params):
--> 151         return cls(**dict(zip(cls.arg_constraints.keys(), params)))
    152 
    153     @staticmethod

/opt/conda/miniconda3/lib/python3.10/site-packages/numpyro/distributions/distribution.py in __call__(cls, *args, **kwargs)
     97             if result is not None:
     98                 return result
---> 99         return super().__call__(*args, **kwargs)
    100 
    101 

/opt/conda/miniconda3/lib/python3.10/site-packages/numpyro/distributions/continuous.py in __init__(self, concentration1, concentration0, validate_args)
    164             jnp.shape(concentration1), jnp.shape(concentration0)
    165         )
--> 166         concentration1 = jnp.broadcast_to(concentration1, batch_shape)
    167         concentration0 = jnp.broadcast_to(concentration0, batch_shape)
    168         super(Beta, self).__init__(batch_shape=batch_shape, validate_args=validate_args)

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py in broadcast_to(array, shape)
   1225 """)
   1226 def broadcast_to(array: ArrayLike, shape: DimSize | Shape) -> Array:
-> 1227   return util._broadcast_to(array, shape)
   1228 
   1229 

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/numpy/util.py in _broadcast_to(arr, shape)
    411 def _broadcast_to(arr: ArrayLike, shape: DimSize | Shape) -> Array:
    412   check_arraylike("broadcast_to", arr)
--> 413   arr = arr if isinstance(arr, Array) else lax.asarray(arr)
    414   if not isinstance(shape, tuple) and np.ndim(shape) == 0:
    415     shape = (shape,)

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/lax/lax.py in asarray(x)
    133   if isinstance(x, np.ndarray) or np.isscalar(x):
    134     # Call device_put_impl directly to avoid binding the primitive.
--> 135     return dispatch._device_put_impl(x)
    136   else:
    137     raise TypeError(f"asarray: expected ArrayLike, got {x} of type {type(x)}.")

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/dispatch.py in _device_put_impl(x, device, src)
    390 
    391   try:
--> 392     aval = xla.abstractify(x)
    393   except TypeError as err:
    394     raise TypeError(

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/interpreters/xla.py in abstractify(x)
    170   typ = type(x)
    171   aval_fn = pytype_aval_mappings.get(typ)
--> 172   if aval_fn: return aval_fn(x)
    173   for typ in typ.__mro__:
    174     aval_fn = pytype_aval_mappings.get(typ)

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/interpreters/xla.py in _make_abstract_python_scalar(typ, val)
    181   # Note: all python scalar types are weak except bool, because bool only
    182   # comes in a single width.
--> 183   return ShapedArray((), dtypes._scalar_type_to_dtype(typ, val),
    184                      weak_type=typ is not bool)
    185 

/opt/conda/miniconda3/lib/python3.10/site-packages/jax/_src/dtypes.py in _scalar_type_to_dtype(typ, value)
    289   if typ is int and value is not None:
    290     if value < np.iinfo(dtype).min or value > np.iinfo(dtype).max:
--> 291       raise OverflowError(f"Python int {value} too large to convert to {dtype}")
    292   return dtype
    293 

**OverflowError: Python int 140034357386608 too large to convert to int32`**
